### PR TITLE
Define internal message flags in PacketHeader to contain the flags for internal use

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
Current packet header flags in transport are 16 bits and only contains the flags need to be encoded into the packet sent out on the wire.

We need to define more flags associated with a inbound or outbound CHIP message for internal use, such as:
The flag indicates if the peer's group key message counter is not synchronized, this is used by MCSP in secure channel.

The flag indicates if the message buffer should not be freed after sending, this is to tell LwIP stack not free the buffer after sending.

Those flags are used internally and not get encoded to the packet sent over the wire. 

 #### Summary of Changes
Define internal message flags in PacketHeader to contain the flags for internal use.

### CODE REVIEW

File: src/transport/raw/MessageHeader.h [hunk 1]
```diff
@@ -72,33 +72,39 @@ enum class ExFlagValues : uint8_t
...
 
+    // Header flag indicates that the peer's group key message counter is not synchronized.
+    kPeerGroupMsgIdNotSynchronized = 0x00010000,
+
+    // Header flag indicates that the message buffer should not be freed after sending.
+    kRetainedBuffer = 0x00020000,
 };
 
-using Flags   = BitFlags<uint16_t, FlagValues>;
+using Flags   = BitFlags<uint32_t, FlagValues>;
```
Comment (turon) [inline src/transport/raw/MessageHeader.h:96]: Couldn't the internal flags be its own separate field?  
Put another way, why combine these two fields in the implementation?
```
uint16_t mInternalFlags;
unit16_t mFlags;
```

File: src/transport/raw/MessageHeader.h [hunk 2]
```diff
@@ -144,6 +150,12 @@ class PacketHeader
     /** Check if it's a secure session control message. */
     bool IsSecureSessionControlMsg() const { return mFlags.Has(Header::FlagValues::kSecureSessionControlMessage); }
 
+    /** Check if the peer's group key message counter is not synchronized. */
+    bool IsPeerGroupMsgIdNotSynchronized() const { return mFlags.Has(Header::FlagValues::kPeerGroupMsgIdNotSynchronized); }
+
+    /** Check if the message buffer should not be freed after sending. */
+    bool IsRetainedBuffer() const { return mFlags.Has(Header::FlagValues::kRetainedBuffer); }
```
Comment (bzbarsky-apple) [inline src/transport/raw/MessageHeader.h:157]: Where and why would this get used?  We used to have a thing for this, and we removed it, due to lack of use cases and the caller being able to just hold a handle if that's what they want.

**Review by woody-apple — CHANGES_REQUESTED** [review]
Couldn't agree more with @turon - if this is just API between layers, why not codify it as such - do we even need flags here? Seems like some simple parameters may even make this easier?

File: src/transport/raw/MessageHeader.h (hunk 2 above)
Comment (yufengwangca) [reply src/transport/raw/MessageHeader.h:157]: We do. https://github.com/project-chip/connectedhomeip/issues/4372

When retaining a buffer with CRMP, the caller expects the msg to be unmodified. LwIP stack will normally prepend the packet headers as the packet traverses the UDP/IP/netif layers, which normally modifies the packet.

We need to deep-copies msg into a fresh object, and queues that for transmission, leaving the original msg available after return.

Comment (bzbarsky-apple) [reply src/transport/raw/MessageHeader.h:157]: I'll follow up in #4372, but I don't see why that needs this API on the packet header.

File: src/transport/raw/MessageHeader.h (hunk 1 above)
Comment (yufengwangca) [reply src/transport/raw/MessageHeader.h]: Acked

File: src/transport/raw/MessageHeader.h (hunk 2 above)
Comment (yufengwangca) [reply src/transport/raw/MessageHeader.h]: This flag finally needs to be passed to Inet layer
INET_ERROR UDPEndPoint::SendMsg(const IPPacketInfo * pktInfo, System::PacketBufferHandle msg, uint16_t sendFlags)

The buffer is retained in messaging layer, currently, we only have PacketHeader and PayloadHeader passed to transport layer. Passing this flag from Messaging layer to Transport layer via PacketHeader seems the only viable path.

I will post another PR to pass this flag to Inet::UDP for  #4372

**Comment by yufengwangca** [comment]
> Couldn't agree more with @turon - if this is just API between layers, why not codify it as such - do we even need flags here? Seems like some simple parameters may even make this easier?

We already have PacketHeader to pass all flags from Messaging layer to Transport layer.

**Review by woody-apple — CHANGES_REQUESTED** [review]
We should not be using flags here. Can we either use a struct with 1 bit members that we need, and add to it, or just have 1 bit members that we add to over time?

**Comment by yufengwangca** [comment]
> We should not be using flags here. Can we either use a struct with 1 bit members that we need, and add to it, or just have 1 bit members that we add to over time?

It is doable, but this struct will be similar to class BitFlags. The two internal flags defined in this PR are immediately needed now, more need to be added. Here is the full set of message flags defined in messaging layer: https://github.com/project-chip/connectedhomeip/blob/master/src/messaging/Flags.h

/**
 *  @brief
 *    Flags associated with a inbound or outbound CHIP message.
 *
 *    The values defined here are for use within the ChipMessageInfo.Flags field.
 */
enum class MessageFlagValues : uint32_t
{
    /**< Indicates that the existing source node identifier must be reused. */
    kReuseSourceId = 0x00000020,
    /**< Indicates that the CHIP message is already encoded. */
    kMessageEncoded = 0x00001000,
    /**< Indicates that default IPv6 source address selection should be used when sending IPv6 multicast messages. */
    kDefaultMulticastSourceAddress = 0x00002000,
    /**< Indicates that the sender of the  message requested an acknowledgment. */
    kPeerRequestedAck = 0x00004000,
    /**< Indicates that the message is a duplicate of a previously received message. */
    kDuplicateMessage = 0x00008000,
    /**< Indicates that the peer's group key message counter is not synchronized. */
    kPeerGroupMsgIdNotSynchronized = 0x00010000,
    /**< Indicates that the source of the message is the initiator of the CHIP exchange. */
    kFromInitiator = 0x00020000,
    /**< Indicates that message is being sent/received via the local ephemeral UDP port. */
    kViaEphemeralUDPPort = 0x00040000,
};

enum class SendMessageFlags : uint16_t
{
    kNone = 0x0000,
    /**< Used to indicate that automatic retransmission is enabled. */
    kAutoRetrans = 0x0001,
    /**< Used to indicate that a response is expected within a specified timeout. */
    kExpectResponse = 0x0002,
    /**< Used to indicate that the source node ID in the message header can be reused. */
    kReuseSourceId = 0x0020,
    /**< Used to indicate that the message is already encoded. */
    kAlreadyEncoded = 0x0080,
    /**< Used to indicate that default IPv6 source address selection should be used when sending IPv6 multicast messages. */
    kDefaultMulticastSourceAddress = 0x0100,
    /**< Used to indicate that the current message is the initiator of the exchange. */
    kFromInitiator = 0x0200,
    /**< Used to send a ReliableMessageProtocol message requesting an acknowledgment. */
    kRequestAck = 0x0400,
    /**< Suppress the auto-request acknowledgment feature when sending a message. */
    kNoAutoRequestAck = 0x0800,
};

Some of them are inherited from Weave and we might not need it eventually, but most of them will be passed to transport layer as corresponding feature comes in.

**Comment by yufengwangca** [comment]
> > We should not be using flags here. Can we either use a struct with 1 bit members that we need, and add to it, or just have 1 bit members that we add to over time?
> 
> It is doable, but this struct will be similar to class BitFlags. The two internal flags defined in this PR are immediately needed now, more need to be added. Here is the full set of message flags defined in messaging layer: https://github.com/project-chip/connectedhomeip/blob/master/src/messaging/Flags.h
> 
> /**
> 
> * @brief
> * Flags associated with a inbound or outbound CHIP message.
> * The values defined here are for use within the ChipMessageInfo.Flags field.
>   */
>   enum class MessageFlagValues : uint32_t
>   {
>   /**< Indicates that the existing source node identifier must be reused. */
>   kReuseSourceId = 0x00000020,
>   /**< Indicates that the CHIP message is already encoded. */
>   kMessageEncoded = 0x00001000,
>   /**< Indicates that default IPv6 source address selection should be used when sending IPv6 multicast messages. */
>   kDefaultMulticastSourceAddress = 0x00002000,
>   /**< Indicates that the sender of the  message requested an acknowledgment. */
>   kPeerRequestedAck = 0x00004000,
>   /**< Indicates that the message is a duplicate of a previously received message. */
>   kDuplicateMessage = 0x00008000,
>   /**< Indicates that the peer's group key message counter is not synchronized. */
>   kPeerGroupMsgIdNotSynchronized = 0x00010000,
>   /**< Indicates that the source of the message is the initiator of the CHIP exchange. */
>   kFromInitiator = 0x00020000,
>   /**< Indicates that message is being sent/received via the local ephemeral UDP port. */
>   kViaEphemeralUDPPort = 0x00040000,
>   };
> 
> enum class SendMessageFlags : uint16_t
> {
> kNone = 0x0000,
> /**< Used to indicate that automatic retransmission is enabled. */ kAutoRetrans = 0x0001, /**< Used to indicate that a response is expected within a specified timeout. */
> kExpectResponse = 0x0002,
> /**< Used to indicate that the source node ID in the message header can be reused. */ kReuseSourceId = 0x0020, /**< Used to indicate that the message is already encoded. */
> kAlreadyEncoded = 0x0080,
> /**< Used to indicate that default IPv6 source address selection should be used when sending IPv6 multicast messages. */ kDefaultMulticastSourceAddress = 0x0100, /**< Used to indicate that the current message is the initiator of the exchange. */
> kFromInitiator = 0x0200,
> /**< Used to send a ReliableMessageProtocol message requesting an acknowledgment. */ kRequestAck = 0x0400, /**< Suppress the auto-request acknowledgment feature when sending a message. */
> kNoAutoRequestAck = 0x0800,
> };
> 
> Some of them are inherited from Weave and we might not need it eventually, but most of them will be passed to transport layer as corresponding feature comes in.

@woody-apple 
I was wondering if we need to invent another struct to represents internal message flags since we already have construct BitFlags.
Right now, we have three sets of flags, one for Exchange Flags, one for Internal Flags and one for Flags need to be encoded into the packet. If we should handle them consistently using the same construct?

**Review by woody-apple — CHANGES_REQUESTED** [review]
Having internal flags that's uint16_t, plus another member that's uint16_t, plus another uint8_t seems super wasteful memory structure wise. I really suggest using bits for these, instead of a uint16_t, given it will save space, but also be more reasonable.

**Comment by yufengwangca** [comment]
> Having internal flags that's uint16_t, plus another member that's uint16_t, plus another uint8_t seems super wasteful memory structure wise. I really suggest using bits for these, instead of a uint16_t, given it will save space, but also be more reasonable.

Hi Woody,

We already use bits for those feature flags, BitFlags is manipulating on bit directly.  Here is the current allocations:
1. ExFlagValues : uint8_t, those bits are defined by the spec for Exchange flags
2. FlagValues : uint16_t, those bits are also defined by the spec, 8 bits for Message flags and 8 bits for Security flags
3. InternalFlagValues : uint16_t those bits are reserved for internal feature flags, currently we only defined two flags in Transport layer, but we have more than 8 internal feature flags defined in messaging layer, we might need to add them to InternalFlagValues as the feature comes in, so I reserved 16 bits here for them.

**Pull request closed without merging**